File: 4 - Replace Conditional Logic with Strategy/C#/MPG.ReplaceConditionalLogicWithStrategy/MPG.ReplaceConditionalLogicWithStrategy.Before/CapitalStrategyTermLoan.cs
using System.Linq;

namespace MPG.ReplaceConditionalLogicWithStrategy.Before
{
    public class CapitalStrategyTermLoan : CapitalStrategy
    {
        public override double Capital(Loan loan)
        {
            return loan.GetCommitment() * loan.Duration() * RiskFactor(loan);
        }

        public override double Duration(Loan loan)
        {
            return WeightedAverageDuration(loan);
        }

        private double WeightedAverageDuration(Loan loan)
        {
            var duration = 0.0;
            var weightedAverage = loan.GetPayments().Sum(payment => YearsTo(payment.Date, loan) * payment.Amount);
            var sumOfPayments = loan.GetPayments().Sum(payment => payment.Amount);

            if (loan.GetCommitment() != 0.0)
            {
                duration = weightedAverage / sumOfPayments;
            }

            return duration;
        }
    }
}

File: 5 - Form Template Method/Python/before/capital_strategy_term_loan.py
from capital_strategy import CapitalStrategy


class CapitalStrategyTermLoan(CapitalStrategy):
    def __init__(self):
        CapitalStrategy.__init__(self)

    def capital(self, loan):
        return loan.get_commitment() * self.duration(loan) * self.risk_factor_for(loan)

    def duration(self, loan):
        return self.weighted_average_duration(loan)

    def weighted_average_duration(self, loan):
        duration = 0.0
        weighted_average = 0.0
        sum_of_payments = 0.0

        for payment in loan.get_payments():
            weighted_average += self.years_to(payment.date, loan) * payment.amount
            sum_of_payments += payment.amount

        if loan.get_commitment() != 0.0:
            duration = weighted_average / sum_of_payments

        return duration
